src/yunohost/data_migrations/0021_migrate_to_bullseye.py

```diff
@@ -30,6 +30,59 @@
 N_NEXT_DEBAN = 11
 N_NEXT_YUNOHOST = 11
 
+VENV_BACKUP_SUFFIX = "_BACKUP_VENV"
+VENV_REQUIREMENTS_SUFFIX = "_req.txt"
+VENV_IGNORE = "ynh_migration_no_regen"
+
+def _get_all_venvs(dir,level=0,maxlevel=2):
+    result = []
+    for file in os.listdir(dir):
+        path = os.path.join(dir,file)
+        if os.path.isdir(path):
+            if os.path.isfile(os.path.join(path,VENV_IGNORE)):
+                continue
+            activatepath = os.path.join(path,"bin","activate")
+            if os.path.isfile(activatepath):
+                content = read_file(activatepath)
+                if "VIRTUAL_ENV" and "PYTHONHOME" in content:
+                    result.append(path)
+                    continue
+            if level<maxlevel:
+                result+=_get_all_venvs(path,level=level+1)
+    return result
+
+def _generate_requirements():
+
+    venvs = _get_all_venvs("/opt/")+_get_all_venvs("/var/www/")
```

Comment:
And what about /opt/yunohost/APP/venv ? I am not sure it works with a max_level =2

Reply:
Fixed in https://github.com/YunoHost/yunohost/pull/1479/commits/b7b4dbfcdff2264011e1b247991bf15a2587a362